File: Windows/tests/WiFiHealthConsole.Core.Tests/ChannelAnalysisTests.cs
using WiFiHealthConsole.Core;

namespace WiFiHealthConsole.Core.Tests;

public sealed class ChannelAnalysisTests
{
    [Fact]
    public void DomainsAndFootprintsCoverTwoPointFourFiveAndSixGHz()
    {
        Assert.Equal((1d, 14d), ChannelAnalysis.Domain(WiFiBand.Band2_4GHz));
        Assert.Equal((32d, 177d), ChannelAnalysis.Domain(WiFiBand.Band5GHz));
        Assert.Equal((1d, 233d), ChannelAnalysis.Domain(WiFiBand.Band6GHz));

        var two = ChannelAnalysis.EstimatedRange(6, 20, WiFiBand.Band2_4GHz);
        var five = ChannelAnalysis.EstimatedRange(149, 80, WiFiBand.Band5GHz);
        var six = ChannelAnalysis.EstimatedRange(229, 160, WiFiBand.Band6GHz);

        Assert.True(two.WidthKnown);
        Assert.Equal(20, two.EffectiveWidthMHz);
        Assert.InRange(five.LowerBound, 32, 177);
        Assert.InRange(five.UpperBound, 32, 177);
        Assert.InRange(six.LowerBound, 1, 233);
        Assert.InRange(six.UpperBound, 1, 233);
    }

    [Fact]
    public void UnknownWidthIsExplicitAndUsesOnlyMinimumVisualizationFootprint()
    {
        var footprint = ChannelAnalysis.EstimatedRange(44, null, WiFiBand.Band5GHz);

        Assert.False(footprint.WidthKnown);
        Assert.Equal(20, footprint.EffectiveWidthMHz);

        var current = Snapshot(WiFiBand.Band5GHz, 44, width: null);
        var suggestion = ChannelAnalysis.Suggestions(current, [Network("far", WiFiBand.Band5GHz, 149, null, -50)])
            .First(item => item.Category == ChannelSuggestionCategory.Channel);

        Assert.Equal(HealthGrade.Unavailable, suggestion.Grade);
        Assert.Equal("参考", suggestion.StatusLabel);
        Assert.Contains("频宽未知", suggestion.Detail);
    }

    [Fact]
    public void OverlapOnlyCountsSameBandAndExcludesCurrentBssid()
    {
        var current = Snapshot(WiFiBand.Band5GHz, 40, 80, bssid: "current");
        var networks = new[]
        {
            Network("current", WiFiBand.Band5GHz, 40, 80, -40),
            Network("near", WiFiBand.Band5GHz, 44, 80, -52),
            Network("far", WiFiBand.Band5GHz, 149, 80, -52),
            Network("other-band", WiFiBand.Band2_4GHz, 6, 20, -52),
        };

        Assert.Equal(1, ChannelAnalysis.OverlapCount(current, networks));
        Assert.True(ChannelAnalysis.OverlapScore(current, networks) > 0);
    }

    [Fact]
    public void StrongerOverlapContributesMoreToScore()
    {
        var current = Snapshot(WiFiBand.Band5GHz, 40, 80);
        var strong = ChannelAnalysis.OverlapScore(current, [Network("strong", WiFiBand.Band5GHz, 44, 80, -45)]);
        var weak = ChannelAnalysis.OverlapScore(current, [Network("weak", WiFiBand.Band5GHz, 44, 80, -85)]);

        Assert.True(strong > weak);
    }

    [Fact]
    public void TwoPointFourFortyMHzRecommendsTwentyAsWarning()
    {
        var current = Snapshot(WiFiBand.Band2_4GHz, 6, 40);

        Assert.Equal(20, ChannelAnalysis.RecommendedWidthMHz(current, []));
        var suggestion = BandwidthSuggestion(current, []);
        Assert.Equal(HealthGrade.Warning, suggestion.Grade);
        Assert.Equal("注意", suggestion.StatusLabel);
        Assert.Equal("建议频宽：20 MHz", suggestion.Title);
    }

    [Fact]
    public void FiveGHzOneHundredSixtyRecommendsEightyInOrdinaryEnvironment()
    {
        var current = Snapshot(WiFiBand.Band5GHz, 40, 160);
        var networks = new[] { Network("distant", WiFiBand.Band5GHz, 149, 80, -55) };

        Assert.Equal(80, ChannelAnalysis.RecommendedWidthMHz(current, networks));
        var suggestion = BandwidthSuggestion(current, networks);
        Assert.Equal(HealthGrade.Warning, suggestion.Grade);
        Assert.Equal("注意", suggestion.StatusLabel);
        Assert.Equal("建议频宽：80 MHz", suggestion.Title);
    }

    [Fact]
    public void FourStrongOverlapsAtEightyRecommendFortyMHz()
    {
        var current = Snapshot(WiFiBand.Band5GHz, 40, 80);
        var networks = new[]
        {
            Network("a", WiFiBand.Band5GHz, 36, 20, -60),
            Network("b", WiFiBand.Band5GHz, 40, 20, -61),
            Network("c", WiFiBand.Band5GHz, 44, 20, -62),
            Network("d", WiFiBand.Band5GHz, 48, 20, -63),
        };

        Assert.Equal(4, ChannelAnalysis.OverlapCount(current, networks));
        Assert.Equal(40, ChannelAnalysis.RecommendedWidthMHz(current, networks));
        Assert.Equal("建议频宽：40 MHz", BandwidthSuggestion(current, networks).Title);
    }

    [Fact]
    public void SameSsidStrongFiveGHzCanBeRecommendedFromTwoPointFour()
    {
        var current = Snapshot(WiFiBand.Band2_4GHz, 6, 20, ssid: "Home");
        var networks = new[]
        {
            Network("home-5", WiFiBand.Band5GHz, 36, 80, -55, ssid: "Home"),
        };

        var suggestion = BandSuggestion(current, networks);
        Assert.Equal("推荐频段：5 GHz", suggestion.Title);
        Assert.Equal(HealthGrade.Unavailable, suggestion.Grade);
        Assert.Equal("参考", suggestion.StatusLabel);
    }

    [Fact]
    public void DifferentSsidNeverTriggersBandSwitch()
    {
        var current = Snapshot(WiFiBand.Band2_4GHz, 6, 20, ssid: "Home");
        var networks = new[]
        {
            Network("guest-5", WiFiBand.Band5GHz, 36, 80, -35, ssid: "Guest"),
            Network("office-6", WiFiBand.Band6GHz, 37, 80, -30, ssid: "Office"),
        };

        var suggestion = BandSuggestion(current, networks);
        Assert.Equal("保持频段：2.4 GHz", suggestion.Title);
        Assert.Equal(HealthGrade.Good, suggestion.Grade);
        Assert.Equal("正常", suggestion.StatusLabel);
    }

    [Fact]
    public void EverySuggestionSetContainsBandBandwidthAndChannel()
    {
        var scenarios = new (WifiSnapshot Current, NearbyNetwork[] Networks)[]
        {
            (WifiSnapshot.Unavailable, []),
            (Snapshot(WiFiBand.Band2_4GHz, 6, 40), []),
            (Snapshot(WiFiBand.Band5GHz, 40, 80), [Network("neighbor", WiFiBand.Band5GHz, 44, 80, -60)]),
            (Snapshot(WiFiBand.Band6GHz, 37, null), [Network("six", WiFiBand.Band6GHz, 69, null, -65)]),
        };

        foreach (var scenario in scenarios)
        {
            var categories = ChannelAnalysis.Suggestions(scenario.Current, scenario.Networks)
                .Select(suggestion => suggestion.Category)
                .ToHashSet();
            Assert.Equal(
                new HashSet<ChannelSuggestionCategory>(Enum.GetValues<ChannelSuggestionCategory>()),
                categories);
        }
    }

    [Fact]
    public void ScanOverlapNeverClaimsCriticalWithoutCcaOrRetransmissionEvidence()
    {
        var current = Snapshot(WiFiBand.Band5GHz, 40, 160);
        var networks = Enumerable.Range(1, 6)
            .Select(index => Network($"n{index}", WiFiBand.Band5GHz, 36 + index * 4, 80, -45 - index))
            .ToArray();

        var suggestions = ChannelAnalysis.Suggestions(current, networks);
        Assert.NotEmpty(suggestions);
        Assert.DoesNotContain(suggestions, suggestion => suggestion.Grade == HealthGrade.Critical);
        Assert.All(suggestions, suggestion =>
            Assert.True(HealthStandards.IsStatusLabelCompatible(suggestion.StatusLabel, suggestion.Grade)));
    }

    [Fact]
    public void CandidateRankingSupportsSixGHzAndUnknownNeighborWidth()
    {
        var current = Snapshot(WiFiBand.Band6GHz, 37, 80);
        var candidates = ChannelAnalysis.ScoreCandidateChannels(
            current,
            [Network("unknown-width", WiFiBand.Band6GHz, 37, null, -50)]);

        Assert.Equal(8, candidates.Count);
        Assert.Contains(candidates, candidate => candidate.PrimaryChannel == 229);
        Assert.Contains(candidates, candidate => candidate.HasUnknownWidthEvidence);
        Assert.Equal(candidates.Min(candidate => candidate.Score), candidates[0].Score);
    }

    private static ChannelSuggestion BandSuggestion(WifiSnapshot current, IEnumerable<NearbyNetwork> networks) =>
        ChannelAnalysis.Suggestions(current, networks)
            .First(suggestion => suggestion.Category == ChannelSuggestionCategory.Band);

    private static ChannelSuggestion BandwidthSuggestion(WifiSnapshot current, IEnumerable<NearbyNetwork> networks) =>
        ChannelAnalysis.Suggestions(current, networks)
            .First(suggestion => suggestion.Category == ChannelSuggestionCategory.Bandwidth);

    private static WifiSnapshot Snapshot(
        WiFiBand band,
        int channel,
        int? width,
        int rssi = -50,
        string ssid = "Home",
        string bssid = "current") => new()
        {
            IsConnected = true,
            Ssid = Observed<string>.Available(ssid, EvidenceSource.Mock),
            Bssid = Observed<string>.Available(bssid, EvidenceSource.Mock),
            Band = Observed<WiFiBand>.Available(band, EvidenceSource.Mock),
            PrimaryChannel = Observed<int>.Available(channel, EvidenceSource.Mock),
            ChannelWidthMHz = width is > 0
            ? Observed<int>.Available(width.Value, EvidenceSource.Mock)
            : Observed<int>.Unavailable(MetricAvailability.Unavailable, EvidenceSource.Mock, "频宽未知"),
            RssiDbm = Observed<int>.Available(rssi, EvidenceSource.Mock),
        };

    private static NearbyNetwork Network(
        string id,
        WiFiBand band,
        int channel,
        int? width,
        int rssi,
        string? ssid = null) => new()
        {
            Id = id,
            Ssid = ssid ?? id,
            Bssid = id,
            Band = band,
            PrimaryChannel = channel,
            ChannelWidthMHz = width,
            WidthEstimated = false,
            RssiDbm = rssi,
            SignalQualityPercent = 70,
        };
}

File: Tests/WiFiHealthConsoleTests/ChannelAnalysisTests.swift
import XCTest
@testable import WiFiHealthConsole

final class ChannelAnalysisTests: XCTestCase {
    func testOverlapCountsNetworksInSameBand() {
        let current = currentSnapshot(band: .band5, channel: 40, width: 80)
        let networks = [
            network(id: "near", band: .band5, channel: 44, width: 80, rssi: -52),
            network(id: "far", band: .band5, channel: 149, width: 80, rssi: -52),
            network(id: "other-band", band: .band2, channel: 6, width: 20, rssi: -52)
        ]

        XCTAssertEqual(ChannelAnalysis.overlapCount(current: current, networks: networks), 1)
    }

    func testEverySuggestionSetContainsBandBandwidthAndChannelCategories() {
        let scenarios: [(WiFiSnapshot, [NearbyNetwork])] = [
            (.unavailable, []),
            (currentSnapshot(band: .band2, channel: 6, width: 40), []),
            (
                currentSnapshot(band: .band5, channel: 40, width: 80),
                [network(id: "neighbor", band: .band5, channel: 44, width: 80, rssi: -60)]
            )
        ]

        for (current, networks) in scenarios {
            let suggestions = ChannelAnalysis.suggestions(current: current, networks: networks)
            XCTAssertEqual(
                Set(suggestions.map(\.category)),
                Set(ChannelSuggestionCategory.allCases),
                "每组建议都必须覆盖频段、频宽和信道"
            )
        }
    }

    func testTwoPointFourGHzFortyMHzRecommendsTwentyMHzAsWarning() throws {
        let current = currentSnapshot(band: .band2, channel: 6, width: 40)

        XCTAssertEqual(ChannelAnalysis.recommendedWidthMHz(current: current, networks: []), 20)

        let suggestion = try bandwidthSuggestion(current: current, networks: [])
        XCTAssertEqual(suggestion.grade, .warning)
        XCTAssertEqual(suggestion.statusLabel, "注意")
        XCTAssertEqual(suggestion.title, "建议频宽：20 MHz")
    }

    func testStrongSameSSIDFiveGHzNetworkIsRecommendedFromTwoPointFourGHz() throws {
        let current = currentSnapshot(ssid: "Home", band: .band2, channel: 6, width: 20)
        let networks = [
            network(
                id: "home-5g",
                ssid: "Home",
                band: .band5,
                channel: 36,
                width: 80,
                rssi: -55
            )
        ]

        let suggestion = try bandSuggestion(current: current, networks: networks)
        XCTAssertEqual(suggestion.title, "推荐频段：5 GHz")
        XCTAssertEqual(suggestion.grade, .unavailable)
        XCTAssertEqual(suggestion.statusLabel, "参考")
    }

    func testFiveGHzOneHundredSixtyMHzRecommendsEightyMHzInOrdinaryEnvironment() throws {
        let current = currentSnapshot(band: .band5, channel: 40, width: 160)
        let networks = [
            network(id: "distant-neighbor", band: .band5, channel: 149, width: 80, rssi: -55)
        ]

        XCTAssertEqual(ChannelAnalysis.recommendedWidthMHz(current: current, networks: networks), 80)

        let suggestion = try bandwidthSuggestion(current: current, networks: networks)
        XCTAssertEqual(suggestion.grade, .warning)
        XCTAssertEqual(suggestion.statusLabel, "注意")
        XCTAssertEqual(suggestion.title, "建议频宽：80 MHz")
    }

    func testFourStrongOverlapsInsideEightyMHzRangeRecommendFortyMHz() throws {
        let current = currentSnapshot(band: .band5, channel: 40, width: 80)
        let networks = [36, 40, 44, 48].enumerated().map { index, channel in
            network(
                id: "strong-overlap-\(index)",
                band: .band5,
                channel: channel,
                width: 20,
                rssi: -60 - index
            )
        }

        XCTAssertEqual(ChannelAnalysis.overlapCount(current: current, networks: networks), 4)
        XCTAssertEqual(ChannelAnalysis.recommendedWidthMHz(current: current, networks: networks), 40)

        let suggestion = try bandwidthSuggestion(current: current, networks: networks)
        XCTAssertEqual(suggestion.grade, .warning)
        XCTAssertEqual(suggestion.statusLabel, "注意")
        XCTAssertEqual(suggestion.title, "建议频宽：40 MHz")
    }

    func testStrongHigherBandNetworksWithDifferentSSIDDoNotTriggerBandSwitch() throws {
        let current = currentSnapshot(ssid: "Home", band: .band2, channel: 6, width: 20)
        let networks = [
            network(id: "guest-5g", ssid: "Guest", band: .band5, channel: 36, width: 80, rssi: -35),
            network(id: "guest-6g", ssid: "Office", band: .band6, channel: 37, width: 80, rssi: -30)
        ]

        let suggestion = try bandSuggestion(current: current, networks: networks)
        XCTAssertEqual(suggestion.title, "保持频段：2.4 GHz")
        XCTAssertEqual(suggestion.grade, .good)
        XCTAssertEqual(suggestion.statusLabel, "正常")
    }

    func testAllSuggestionStatusLabelsRemainCompatibleWithGrades() {
        let crowdedNetworks = [36, 40, 44, 48].enumerated().map { index, channel in
            network(
                id: "crowded-\(index)",
                band: .band5,
                channel: channel,
                width: 20,
                rssi: -60 - index
            )
        }
        let scenarios: [(WiFiSnapshot, [NearbyNetwork])] = [
            (.unavailable, []),
            (currentSnapshot(band: .band2, channel: 6, width: 40), []),
            (
                currentSnapshot(ssid: "Home", band: .band2, channel: 6, width: 20),
                [network(id: "home-5g", ssid: "Home", band: .band5, channel: 36, width: 80, rssi: -55)]
            ),
            (currentSnapshot(band: .band5, channel: 40, width: 160), []),
            (currentSnapshot(band: .band5, channel: 40, width: 80), crowdedNetworks),
            (
                currentSnapshot(ssid: "Home", band: .band2, channel: 6, width: 20),
                [network(id: "other-5g", ssid: "Other", band: .band5, channel: 36, width: 80, rssi: -30)]
            )
        ]

        for (current, networks) in scenarios {
            for suggestion in ChannelAnalysis.suggestions(current: current, networks: networks) {
                XCTAssertTrue(
                    HealthStandards.isStatusLabelCompatible(suggestion.statusLabel, with: suggestion.grade),
                    "状态文字 \(suggestion.statusLabel) 与等级 \(suggestion.grade) 不兼容：\(suggestion.title)"
                )
            }
        }
    }

    func testOverlapSuggestionsNeverBecomeCriticalWithoutCCAOrRetransmissionEvidence() {
        let current = currentSnapshot(band: .band5, channel: 40, width: 160)
        let networks = (1...6).map { index in
            network(
                id: "neighbor-\(index)",
                band: .band5,
                channel: 36 + index * 4,
                width: 80,
                rssi: -45 - index
            )
        }

        let suggestions = ChannelAnalysis.suggestions(current: current, networks: networks)
        XCTAssertFalse(suggestions.isEmpty)
        XCTAssertFalse(
            suggestions.contains(where: { $0.grade == .critical }),
            "扫描重叠不能脱离 CCA/重传单独判严重"
        )
    }

    private func bandSuggestion(
        current: WiFiSnapshot,
        networks: [NearbyNetwork],
        file: StaticString = #filePath,
        line: UInt = #line
    ) throws -> ChannelSuggestion {
        try XCTUnwrap(
            ChannelAnalysis.suggestions(current: current, networks: networks)
                .first(where: { $0.category == .band }),
            "缺少频段建议",
            file: file,
            line: line
        )
    }

    private func bandwidthSuggestion(
        current: WiFiSnapshot,
        networks: [NearbyNetwork],
        file: StaticString = #filePath,
        line: UInt = #line
    ) throws -> ChannelSuggestion {
        try XCTUnwrap(
            ChannelAnalysis.suggestions(current: current, networks: networks)
                .first(where: { $0.category == .bandwidth }),
            "缺少频宽建议",
            file: file,
            line: line
        )
    }

    private func currentSnapshot(
        ssid: String = "Home",
        band: WiFiBand,
        channel: Int,
        width: Int,
        rssi: Int = -50
    ) -> WiFiSnapshot {
        WiFiSnapshot(
            ssid: ssid,
            bssid: "current",
            band: band,
            channel: channel,
            channelWidthMHz: width,
            rssi: rssi,
            noise: -90
        )
    }

    private func network(
        id: String,
        ssid: String? = nil,
        band: WiFiBand,
        channel: Int,
        width: Int,
        rssi: Int
    ) -> NearbyNetwork {
        NearbyNetwork(
            id: id,
            ssid: ssid ?? id,
            bssid: id,
            band: band,
            channel: channel,
            channelWidthMHz: width,
            rssi: rssi,
            noise: -90
        )
    }
}
